qemu/tests/thin_provisioning.py

```diff
@@ -0,0 +1,176 @@
+        logging.debug("Block allocation bitmap: %s" % get_allocation_bitmap())
+        raise exceptions.TestError(
+            "Block allocation bitmap not empty before test.")
+
+    vm_name = params["main_vm"]
+    params["start_vm"] = "yes"
+    params["image_name_stg"] = disk_name
+    error_context.context(
+        "TEST STEP 2: Boot guest with disk {:s}".format(disk_name), logging.info)
+    env_process.preprocess_vm(test, params, env, vm_name)
+    vm = env.get_vm(vm_name)
+    vm.verify_alive()
+    timeout = float(params.get("login_timeout", 240))
+    session = vm.wait_for_login(timeout=timeout)
+
+    guest_device_name = get_guest_discard_disk(session)
+    rewrite_disk_cmd = params["rewrite_disk_cmd"]
+    rewrite_disk_cmd = rewrite_disk_cmd.replace("DISK", guest_device_name)
+    error_context.context(
+        "TEST STEP 3: Rewrite the disk in {:s}, "
+        "using the command {:s}".format(
+            guest_device_name, rewrite_disk_cmd), logging.info)
+    session.cmd(rewrite_disk_cmd, timeout=timeout, ignore_all_errors=True)
+    bitmap_before_trim = get_allocation_bitmap()
+    error_context.context("TEST STEP 4: Check bitmap before test: {:s}".format(
+        bitmap_before_trim), logging.info)
+    if not re.match(r"\d+-\d+", bitmap_before_trim):
+        raise exceptions.TestFail(
+            "Bitmap should be continuous before fstrim, {}".format(
+                bitmap_before_trim))
+
+    format_disk_cmd = params["format_disk_cmd"]
+    format_disk_cmd = format_disk_cmd.replace("DISK1", guest_device_name)
+    error_context.context("TEST STEP 5: Format disk '{:s}' in guest".format(
+        guest_device_name), logging.info)
+    session.cmd(format_disk_cmd)
+
+    mount_disk_cmd = params["mount_disk_cmd"]
+    mount_disk_cmd = mount_disk_cmd.replace("DISK1", guest_device_name)
+    session.cmd(mount_disk_cmd)
+
+    error_context.context("TEST STEP 6: Execute fstrim in guest", logging.info)
+    fstrim_cmd = params["fstrim_cmd"]
+    o = session.cmd_output(fstrim_cmd, timeout=timeout)
+    if params.get(
+            "drive_format_stg") == "virtio":
+        if not re.search(params["match_str"], o):
+            raise exceptions.TestError(
+                "Command output is wrong,should be: {}.".format(
+                    params["match_str"]))
+
+    bitmap_after_trim = get_allocation_bitmap()
+    error_context.context("TEST STEP 7: Bitmap after test: {:s}".format(
+        bitmap_after_trim), logging.info)
+    if params.get("drive_format_stg") == "scsi-hd":
+        if not re.match(r"\d+-\d+,.*\d+-\d+$", bitmap_after_trim):
+            raise exceptions.TestFail("discard command doesn't issue"
+                                      "to scsi_debug disk, please report "
+                                      "bug for qemu")
+    else:
+        if bitmap_before_trim != bitmap_after_trim:
+            raise exceptions.TestError(
+                "The disk is virtio_blk, the map should "
```

Comment:
i mean only virito_scsi is for the disk used for test, not the os disk. the virtio_blk is used for negative test, which is also have a case and the case name is: Negative test on virtio-blk for thinly-provisioned lun with discard enabled (write same/unmap mode)